Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp

```diff
@@ -0,0 +1,540 @@
+              // Real Hx_eff = Hx_zface + Hx_bias_zface;
+              // Real Hy_eff = Hy_zface + Hy_bias_zface;
+              // Real Hz_eff = Hz_zface + Hz_bias_zface;
+              Real Hx_eff = Hx_bias_zface;
+              Real Hy_eff = Hy_bias_zface;
+              Real Hz_eff = Hz_bias_zface;
+
+              // magnetic material properties mag_alpha and mag_Ms are defined at cell nodes
+              // keep the interpolation
+              Real mag_gamma_interp = MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(0,0,1),mag_gamma_arr);
+
+              Real a_temp_dynamic_coeff = PhysConst::mu0 * std::abs(mag_gamma_interp) *
+                        (1.0 + std::pow(MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(0,0,1),mag_alpha_arr), 2.0))/ 2.0;
+
+              // calculate a_temp_zface
+              // x component on z-faces of grid
+              a_temp_zface(i, j, k, 0) = -( dt * a_temp_dynamic_coeff * Hx_eff + a_temp_static_zface(i, j, k, 0) );
+
+              // y component on z-faces of grid
+              a_temp_zface(i, j, k, 1) = -( dt * a_temp_dynamic_coeff * Hy_eff + a_temp_static_zface(i, j, k, 1) );
+
+              // z component on z-faces of grid
+              a_temp_zface(i, j, k, 2) = -( dt * a_temp_dynamic_coeff * Hz_eff + a_temp_static_zface(i, j, k, 2) );
+
+              // calculate M_zface
+              // x component on z-faces of grid
+              M_zface(i, j, k, 0) = MacroscopicProperties::getM_half (i, j, k, 0, a_temp_zface, b_temp_zface);
+
+              // y component on z-faces of grid
+              M_zface(i, j, k, 1) = MacroscopicProperties::getM_half (i, j, k, 1, a_temp_zface, b_temp_zface);
+
+              // z component on z-faces of grid
+              M_zface(i, j, k, 2) = MacroscopicProperties::getM_half (i, j, k, 2, a_temp_zface, b_temp_zface);
+
+              // x component on z-faces of grid
+              M_error_zface(i, j, k, 0) = std::abs((M_zface(i, j, k, 0) - M_prev_zface(i, j, k, 0))) / (std::abs(M_prev_zface(i, j, k, 0)) + 1.0e-12);
+
+              // y component on z-faces of grid
+              M_error_zface(i, j, k, 1) = std::abs((M_zface(i, j, k, 1) - M_prev_zface(i, j, k, 1))) / (std::abs(M_prev_zface(i, j, k, 1)) + 1.0e-12);
+
+              // z component on z-faces of grid
+              M_error_zface(i, j, k, 2) = std::abs((M_zface(i, j, k, 2) - M_prev_zface(i, j, k, 2))) / (std::abs(M_prev_zface(i, j, k, 2)) + 1.0e-12);
+              });
+        }
+
+        // Check the error between Mfield and Mfield_prev and decide whether another iteration is needed
+        for (int i = 0; i < 1; i++){
+            for (int j = 0; j < 3; j++){
+                Real M_iter_error = Mfield_error[i]->norm0(j);
+                if(M_iter_error >= M_iter_maxerror) {
+                     M_iter_maxerror = M_iter_error;
+                }
+            }
+        }
```

Comment:
Note this is very strict criterion which may lead to many iterations when we turn on H_Maxwell in H_eff...
An alternative way is to calculate the average error, which loosen the requirement.

Reply:
Yes.  If we define "error" as the absolute value of the change from one iteration to the next, that seems better than relative error, which is perilous when M is close to zero or transitioning from positive to negative.  See my comments in slack.